app/apps/server/bridges/uploads.js

```diff
@@ -30,20 +30,31 @@ export class AppUploadBridge {
 		});
 	}
 
-	createUpload(details, buffer, appId) {
+	async createUpload(details = {}, buffer, appId) {
 		this.orch.debugLog(`The App ${ appId } is creating an upload "${ details.name }"`);
 
+		if (!details.userId) {
+			throw new Error('Missing user to perform the upload operation');
+		}
+
 		details.type = determineFileType(buffer, details);
 
-		const fileStore = FileUpload.getStore('Uploads');
-		const uploadedFile = fileStore.insertSync(details, buffer);
+		return new Promise((resolve, reject) => Meteor.runAsUser(details.userId, () => {
+			try {
+				const fileStore = FileUpload.getStore('Uploads');
+				const insertSync = Meteor.wrapAsync(fileStore.insert.bind(fileStore));
+				const uploadedFile = insertSync(details, buffer);
 
-		if (details.userId) {
-			Meteor.runAsUser(details.userId, () => {
-				Meteor.call('sendFileMessage', details.rid, null, uploadedFile);
-			});
-		}
+				if (details.visitorToken) {
+					Meteor.call('sendFileLivechatMessage', details.rid, details.visitorToken, uploadedFile);
+				} else {
+					Meteor.call('sendFileMessage', details.rid, null, uploadedFile);
+				}
 
-		return this.orch.getConverters().get('uploads').convertToApp(uploadedFile);
+				resolve(this.orch.getConverters().get('uploads').convertToApp(uploadedFile));
+			} catch (err) {
+				reject(err);
+			}
+		}));
```

Comment:
As discussed before, `userId` is not available when the upload is made by a livechat visitor, so we can't use `Meteor.runAsUser` here.

Alternatively, we can actually use `Meteor.bindEnvironment`, which should allow us to use `fileStore.insertSync(details, buffer);` without the error faced earlier.

Then, we can call `Meteor.runAsUser` to wrap the call to `Meteor.call('sendFileMessage')`, as it requires `Meteor.userId()` to be valid.